### ISSUE TITLE
meaning of cast `:` and `safeCast` for domains and ranges

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#19269]
This issue asks to define the semantics of cast (operator `:`) from a rectangular domain to a rectangular domain type and whether we need a separate `safeCast`. These decisions will carry over to ranges.

This issue also considers casts between other combinations of domain types.

## Rectangular domains

Here are the things that we may want a cast or "safeCast" to do:

* If the from-domain is stridable and the to-domain is non-stridable:
  - issue a compiler error
  - run-time check that the from-domain's stride == 1, otherwise:
    - halt the program, or
    - throw an exception
    - if this runtime check be performed only when `boundsChecking` or `castChecking` is true, throwing an exception is not appropriate just like it is not for the array bounds check
  - or skip the runtime check and always drop the from-domain's stride

* If the from-domain's idxType is coercible, i.e. convertible implicitly, to the to-domain's idxType:
  - allow such a cast, or
  - issue a compiler error

* If the from-domain's idxType is cast-able to the to-domain's idxType:
  - allow such a cast always
  - allow such a cast with a runtime check, under the same considerations as the stride check above
  - issue a compiler error

The straw-person proposal is to follow the precedent of integer casts:

* the cast operator `:` drops the stride and casts the indices, if needed for a different idxType, without a runtime check
* the `safeCast` method performs a runtime check for the stride and a safeCast() for the indices, calling `HaltWrappers.safeCastCheckHalt()` if the check fails

A straw-person counter-proposal is to have the cast operator `:` do what what the proposal's `safeCast` method does, and avoid defining `safeCast` on domains altogether.

We can initially postpone implementing support for differing `idxType`s and issue a compiler error instead. Such support can be added later upon user interest while maintaining backwards compatibility.

## Irregular domains

Casting from an irregular domain to a (non-sparse) rectangular domain should be a compiler error, as such an operation is not straightforward and unlikely to be sought by the user.

Casting between other combinations of domain types is well defined semantically, however is also unlikely needed by the user. We can have such casts issue a compiler error initially and implement as needed.

When casting between two domains with different `idxType`, we should follow the same strategy that we choose for rectangular domains.

**Comment by lydia-duncan** [comment]
Note that if we make this change, we will have to update the spec section on [explicit range and domain conversions](https://chapel-lang.org/docs/latest/language/spec/conversions.html#explicit-range-conversions)

**Comment by lydia-duncan** [comment]
What I wonder about when considering this change is: what is the strict definition of casts and safeCasts?  I wasn't seeing documentation on safeCast as a concept, but I know it's something we're using in a few cases.  It seems like understanding that is important for making this choice

**Comment by mppf** [comment]
> when we still need to decide whether stridability should be changed from a bool to an enum. 

I think this is issue #17131.

**Comment by vasslitvinov** [comment]
Rust: https://docs.rs/safecast/latest/safecast/
Swift: https://developer.apple.com/swift/blog/?id=23

**Comment by vasslitvinov** [comment]
Notes from today's group meeting: we considered three ways to convert a value to a more specific range type from a more general range type, ex. `range(int(8), unstridable)` <-- `range(int(64), stridable)`
* assignment / initialization `=`
* cast `:`
* safeCast()

**safeCast():** we had diverse opinions, from "not a problem" to "funky and ugly". We leaned towards the latter and easily converged on moving its functionality to cast `:` and removing safeCast().

**Cast `:`:** we disliked the current approach of dropping the stride silently when going from a stridable to an unstridable range. Our users dislike it too. We chose for the cast to throw in this case.

We favored the cast to check that the bounds fit in the target `idxType` and throw if they don't. This is for consistency with the case where the stride does not "fit" in the target type.

We compared this behavior against narrowing casts on integral types, like int(8) <-- int(64), which silently drop bits. We justified doing it differently for ranges because narrowing casts on integers are a well-known operation in other languages. Sometimes they are used intentionally to drop bits. Casts on ranges are not in this "well-known" category so we would be much less justified to drop bits and strides silently. Dropping higher-order bits on range bounds is more likely to be a bug than intentional behavior.

We chose to keep the checks for idxType and the stride enabled regardless of the `--fast` setting. This is our convention for the cases where violation of a runtime check results in throwing, to preserve the throwing behavior regardless of `--fast`. We expect these checks to be needed rarely and so not to be a performance bottleneck. Some of our users support this conclusion, too. This is by contrast to narrowing casts on integral types, which occur within tight loops in some user codes and so could reduce performance if checking were enabled. If user code needs to skip the runtime check when casting ranges, they would write the desired conversion on their own.

**Assignment/initialization to a more specific from a more general range type:** the motivation for allowing it is to have a convenient way to create, for example, positive- or negative-stride ranges being proposed #17131. We decided to keep this kind of assignment illegal because it is illegal in analogous situations with other types. Also the motivation is not strong enough. Indeed, we expect this operation to be infrequent and, when needed, it can be expressed with a cast, which would be only modestly inconvenient, if at all.

**Comment by bradcray** [comment]
> when needed, it can be expressed with a cast, which would be only modestly inconvenient, if at all.

I also think an important tool is to sometimes just not worry about converting the old range in any way and just create a new one from scratch.  E.g.,

```chapel
var r = whatever;
var rNegStride = r.low..r.high by -1;  // or 'by -r.stride' ... or perhaps 'r.lowBound..r.highBound'
```

Given that range type signatures can be somewhat complicated, this way seems to express the code's intent clearly and succinctly and avoids a lot of "Is this cast legal?" kind of soul-searching.  As a user, it's often what I think I'd reach for.

**Comment by vasslitvinov** [comment]
@bradcray, @mppf, all -- what do you think about throwing vs. halting in non-compliant cases? Throwing is more hassle for the user than halting. If we went for halting, we would probably be disabling it upon --fast.

**Comment by mppf** [comment]
I think if we want halting, we should have `safeCast` offer that. I would expect that `:` will throw and not halt.

**Comment by bradcray** [comment]
Since we have other casts that throw, I think throwing is the right answer here.  I also don't think casting ranges is all that common, so don't think that the hassle will be all that big of a deal.

**Comment by vasslitvinov** [comment]
> I also think an important tool is to sometimes just not worry about converting the old range in any way and just create a new one from scratch.

Two ideas on this topic, which can be post-2.0:
* `new range(......)`; the challenge is to make it reasonably succinct in typical cases
* `range.adjustStride(...)` and perhaps other range.tweakThisOrThat() methods; akin to range.expand() et al.

**Comment by vasslitvinov** [comment]
Here are the spots in our code that the above proposal https://github.com/chapel-lang/chapel/issues/19269#issuecomment-1299406718 would affect / disallow:

* Domain cast `:` is used to drop the stride intentionally when calculating the bounding box:

```chpl
// BlockDist; ditto StencilDist
proc Block.init(boundingBox: domain,...) {
  ...
  this.boundingBox = boundingBox : domain(rank, idxType, stridable = false);
  ...
}
```

* Range `safeCast` is used to convert from stridable type to unstridable in cases where:
  * unstridability is known at compile time, yet...
  * the code is generic; it is written using `by` so it works for any range stridability:

```chpl
// BlockDom follower iterator, ditto StencilDom; simpler for BlockCyclicDom
// t(i) is unstridable ==> 'this' and 'followThis' are unstridable
var t: rank*range(idxType, stridable=stridable||anyStridable(followThis));
...
t(i) = ((low..high by stride:strType) + whole.dim(i).alignedLow by followThis(i).stride:strType).safeCast(t(i).type);
```

* DimensionalDist2D casts away range stridability for an unknown reason in the domain leader. My best guess is that it keeps the internal representation stridable for simplicity, then converts it to unstridable before yielding when the user expects that. Or else it intentionally drops the stride, given that happens with leader chunks that are supposed to be dense.
```chpl
iter DimensionalDom.these(param tag: iterKind) where tag == iterKind.leader:
  ....
  yield loc1d.dsiMyDensifiedRangeForTaskID1d
    (dom1d, taskid, numTasks) : resultT;

iter Block1locdom.dsiMyDensifiedRangeForSingleTask1d(globDD):
    ....
    yield locRange: range(globDD.idxType);
```

The are pretty specialized codes. If the proposal introduces inconvenience in these cases, it should be OK. I suggest a method like `domain.boundingBox()` for general user consumption, post-2.0.

**Comment by bradcray** [comment]
None of these particularly give me pause.  Give a shout if you can't find reasonable workarounds to avoid the use of casting.

**Comment by vasslitvinov** [comment]
We revisited this topic in today's group meeting and preferred the following operations:
* the cast operator `:` is to check only the stride and throws if the original stride does not fit in the target type, and
* introduce the method `tryCast` that checks everything and throws if a check fails, i.e, behaves like the cast operator as implemented in #22858

This preference was motivated by reconsidering the cast operator in the light of #22858 and the following considerations:
* We want to allow users to narrow the `idxType` of a range or domain easily, i.e., without worrying about the possibility or an exception being thrown. This is analogous to how casts on integral types work.
* It does not make sense for the cast operator to ignore stride mismatches the way it did before #22858. If the user needs to drop the stride of a range `r`, they should use a range literal like `(r.low..r.high)` or, especially for domains, perhaps a boundingBox() method.
* In a previous group discussion, which we had after #22858, we expressed the preference for the cast operator to avoid checking and throwing when casting from an enum idxType to an integral idxType, eg., from `range(myEnum)` to `range(int(8))`. This is analogous to how the casts between primitive types work, i.e., when casting an enum value to an integral type. It would be inconsistent if casts from an enum idxType to an integral idxType never threw while casts between integral idxTypes could throw in some cases.

**Comment by vasslitvinov** [comment]
We want to apply the proposal of a compiler warning for certain casts from an enum to an integer in https://github.com/chapel-lang/chapel/issues/22842#issuecomment-1670458090 to corresponding casts from range(enum) to range(integral).

See also #22905 for a discussion of casts and tryCast into `range(bool)` .

**Comment by bradcray** [comment]
I'd either have it print a diagnostic warning at execution-time and let the wheels fall off the car where they may, or possibly halt, with a slight preference for the former.

**Comment by bradcray** [comment]
I'd be OK with that, personally.

**Comment by vasslitvinov** [comment]
We recently favored convenience and performance over runtime checking for range casts to a narrower idxType, see https://github.com/chapel-lang/chapel/issues/19269#issuecomment-1678271195 above:
```chpl
// convenient; user must ensure that 'a' and 'b' fit in int(8)
var r = (a..b): range(int(8));
```

I suggest that we also want to favor convenience and performance for range casts to a narrower strideKind:
```chpl
// user must also ensure that 'myStride' fits in int(8) and is positive
var r = (a..b by myStride): range(int(8), strides=strideKind.positive);
```

This approach comes especially handy in code that is generic over a range. For example, our `RangeChunk.chunk(r: range(?), ...)` returns a new range that inherits the stride from `r`, where we want this new range to have the same type as `r`. With my proposal, this is conveniently written with a cast:
```chpl
  proc chunk(r: range(?), ...) {
    .......
    return ( start..end by r.stride ): r.type;
  }
```

Without my proposal, convincing the compiler to accept r.type for `start..end by r.stride` either hits #22575 or requires exception-related runtime overhead:
```chpl
    return try! ( start..end by r.stride ).tryCast(r.type);
```

To address my concern that casting the stride to a narrower type my result in violating range invariants, see https://github.com/chapel-lang/chapel/issues/19269#issuecomment-1683228068 above, I suggest that we check at runtime both that (a) the original stride is acceptable for the `strides` parameter of the target range type and (b) that casting the stride to the target `idxType` preserves the stride. Both checks are to be disabled upon --fast / --no-checks. This is an extension of my suggestion in https://github.com/chapel-lang/chapel/issues/19269#issuecomment-1684293252 and may obviate the need in `chpl_setFields()` in #22575.

**Comment by vasslitvinov** [comment]
The key impacts of my proposal just above "halt if the stride does not fit" are:

(a) To create a positive-stride range/domain, a cast will be natural, for example:
```chpl
// with my proposal
config const step = 2;
var d2 = {1..n by step}: domain(1, strides=strideKind.positive);

// without my proposal
config const step = 2;
var d2 = {1..n by step: uint};   // shorter, yet less intuitive
```

(b) With my proposal, we can make the following change in the standard module `RangeChunks`, which would not be possible without it:
```diff
   proc chunk(r: range(?), ........) {
     const (startOrder, endOrder) = chunkOrder(r, numChunks, idx, remPol);
     const start = r.orderToIndex(startOrder);
     const end = r.orderToIndex(endOrder);
-    var result: r.type;
-    result.chpl_setFields(start, end, r.stride); // start..end by r.stride
-    return result;
+    return ( start..end by r.stride ): r.type;
   }
```

**Comment by vasslitvinov** [comment]
An internal poll shows that the group is in favor of my most recent proposal above https://github.com/chapel-lang/chapel/issues/19269#issuecomment-1685042219

**Comment by vasslitvinov** [comment]
#23099 implemented the revised decisions for the range cast. Closing this issue.

**Comment by lydia-duncan** [comment]
(you forgot to actually close it)
